File: Leetcode/17_swap_nodes_in_pairs.py
# https://leetcode.com/problems/swap-nodes-in-pairs/

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# 재귀로 풀기
def swapPairs(head: ListNode) -> ListNode:

    if head and head.next:
        temp = head.next

        # 스왑된 값을 받는다
        head.next = swapPairs(temp.next)

        temp.next = head

        return temp

    return head


# 반복으로 풀기
# def swapPairs(head: ListNode) -> ListNode:

#     prev = root = ListNode(None)

#     prev.next = head

#     while head and head.next:

#         temp = head.next
#         head.next = temp.next
#         temp.next = head

#         prev.next = temp

#         head = head.next
#         prev = prev.next.next

#     return root.next

File: Leetcode/17_swap_nodes_in_pairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * 재귀로 풀기
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function (head) {
  // 페어가 있을 때만 스왑이 가능하다.
  if (head && head.next) {
    let temp = head.next;
    head.next = swapPairs(temp.next);
    temp.next = head;

    // 페어의 앞쪽 노드를 리턴한다.
    return temp;
  }

  // 페어가 아닐 땐 바로 헤드 노드를 리턴
  return head;
};

/**
 * 반복문 사용
 * @param {ListNode} head
 * @return {ListNode}
 */
// var swapPairs = function (head) {
//   let prev = new ListNode(0);
//   let root = prev;

//   prev.next = head;

//   while (head && head.next) {
//     let temp = head.next;

//     [head.next, temp.next, prev.next] = [temp.next, head, temp];

//     head = head.next;
//     prev = prev.next.next;
//   }

//   return root.next;
// };
